oak_client/src/attestation.rs

```diff
@@ -0,0 +1,156 @@
+    let certs = rustls::internal::pemfile::certs(&mut cc_reader)
+        .map_err(|error| anyhow!("Couldn't parse TLS certificate: {:?}", error))?;
+    if certs.len() != 1 {
+        return Err(anyhow!("Only a single PEM certificate should be provided"));
+    }
+    config
+        .root_store
+        .add(&certs[0])
+        .context("Couldn't add root certificate")?;
+
+    // Set custom X.509 certificate verifier.
+    config
+        .dangerous()
+        .set_certificate_verifier(std::sync::Arc::new(RemoteAttestationVerifier::new(
+            tee_quote,
+        )));
+
+    Ok(ClientTlsConfig::new().rustls_client_config(config))
+}
+
+/// Custom X.509 certificate verifier that checks TEE quotes in the X.509 TEE extension.
+struct RemoteAttestationVerifier {
+    tee_quote: String,
+}
+
+impl RemoteAttestationVerifier {
+    fn new(tee_quote: &str) -> Self {
+        Self {
+            tee_quote: tee_quote.to_string(),
+        }
+    }
+
+    fn check_tee_quote(&self, extension: &str) -> anyhow::Result<()> {
+        let pattern = r"^Quote:(.*)$";
+        let re = regex::Regex::new(&pattern).expect("Incorrect regular expression");
+        let tee_quotes = re.captures(&extension).context("Couldn't find TEE quote")?;
+
+        // `0` capture group corresponds to the whole pattern, while `1` corresponds to the first
+        // match.
+        // https://docs.rs/regex/1.4.3/regex/struct.Captures.html#method.get
+        if self.tee_quote == tee_quotes.get(1).map_or("", |m| m.as_str()) {
+            Ok(())
+        } else {
+            Err(anyhow!("Incorrect TEE quote"))
+        }
+    }
+}
+
+impl rustls::ServerCertVerifier for RemoteAttestationVerifier {
+    fn verify_server_cert(
+        &self,
+        roots: &rustls::RootCertStore,
+        certs: &[rustls::Certificate],
+        hostname: webpki::DNSNameRef,
+        ocsp: &[u8],
+    ) -> Result<rustls::ServerCertVerified, rustls::TLSError> {
+        if certs.len() != 1 {
+            return Err(rustls::TLSError::General(
+                "Only a single certificate should be provided".to_string(),
+            ));
+        }
+        let certificate = &certs[0].0;
+
+        // Parse X.509 certificate encoded using DER format.
+        // https://tools.ietf.org/html/rfc7468
+        let (_, parsed_certificate) = parse_x509_certificate(&certificate).map_err(|error| {
+            rustls::TLSError::General(format!("Couldn't parse certificate: {}", error))
+        })?;
+
+        // Extract X.509 TEE extension.
+        let extensions = parsed_certificate.extensions();
+        let tee_extension = extensions.get(&TEE_EXTENSION_OID).ok_or_else(|| {
+            rustls::TLSError::General("Couldn't find TEE quote X.509 extension".to_string())
+        })?;
+        let tee_extension_value = parse_der(tee_extension.value)
+            .map_err(|error| {
+                rustls::TLSError::General(format!("Couldn't parse TEE extension: {:?}", error))
```

Comment:
Inconsistent naming "TEE extension" vs "TEE quote X.509 extension" above, could you make things consistent?